### ISSUE TITLE
Sparse scores array

### ISSUE THREAD

**Issue by florian-huber** [matchms/matchms#336]
This issue is to replace and update the former discussion on sparse scores arrays in #60.

In many use cases we have seen so far for matchms, it is *not* about computing and scoring all pair-wise scores between spectra. in particular for larger datasets the aim often is to identify likely matches or related compounds etc. Those scenarios have in common that a large fraction of spectral pairs can be ignored very quickly, either because it gets a score of `0` (or `False` when boolean) for scores such as `MetadataMach` or `PrecursorMzMatch`, or because it gets a spectral similarity score below a set threshold for instance for scores such as `CosineGreedy` etc.

Including sparseness into matchms would benefit computations in two ways:
- Saving memory: Only the scores of interest are stored in a sparse array, the rest is set to `0`. Depending on the fraction of scores of interest this can save a very large part of memory.
- Saving computation time: If multiple scores are combined in a pipeline, users would often only be interest in the computation of a certain score for those pairs which got a none-zero score in a precious score computation. Only computing those pairs can save a lot of computation time.

**Comment by florian-huber** [comment]
As a solution to this I experimented with various sparse matrices (from scipy mostly), but didn't find anything that fully worked for what I had in mind. I hence designed a stacked sparse array --> `StackedSparseScores` which is a variation of  a COO-style sparse matrix. This means it stores values by keeping 3 arrays: `row`, `column` and `value`. The main difference to a scipy-style COO matrix is that (a) it can stack multiple values in the `value` array and (b) it has more sophisticated slicing options. Since its custom-made for matchms it also contains methods tailored for our use cases such as `filter_by_range`.

https://github.com/matchms/matchms/blob/be840309e7927343d4d69b2beb2d47d4d5a7b3f6/matchms/StackedSparseScores.py#L11

Since it is meant to become a key part of matchms I added [extensive tests](https://github.com/matchms/matchms/blob/be840309e7927343d4d69b2beb2d47d4d5a7b3f6/tests/test_stacked_sparse_scores.py). Important to note is also that during development other matchms tests only failed for reasons of slicing or similar, but not because any expected score outcames changed.

A code example which is now also in the docstring is:
```python
import numpy as np
from matchms import StackedSparseScores

scores1 = np.random.random((12, 10))
scores2 = np.random.random((12, 10))
scores_array = StackedSparseScores(12, 10)
scores_array.add_dense_matrix(scores1, "scores_1")
scores_array = scores_array.filter_by_range("scores_1", low=0.5)

# Add second scores and filter
scores_array.add_dense_matrix(scores2, "scores_2")
scores_array = scores_array.filter_by_range("scores_2", low=0.1, high=0.4)

# Scores can be accessed using (limited) slicing capabilities
scores_array[3, 4]  # => scores_1 and scores_2 at position row=3, col=4
scores_array[3, :]  # => tuple with row, col, scores for all entries in row=3
scores_array[:, 2]  # => tuple with row, col, scores for all entries in col=2
scores_array[3, :, 0]  # => tuple with row, col, scores_1 for all entries in row=3
scores_array[3, :, "scores_1"]  # => same as the one before

# Scores can also be converted to a dense numpy array:
scores2_after_filtering = scores_array.to_array("scores_2")
```

**Comment by hechth** [comment]
@florian-huber this looks pretty nice! Just one question: the filter_by_range performs in-place filtering? Is this for performance reasons? this just could be an easy source of error for many users.

**Comment by florian-huber** [comment]
Thanks a lot for the feedback!
And you're right. It should probably better not be in-place filtering. I will check if changing this impacts performance, but I guess it won't.

**Comment by florian-huber** [comment]
I changed the code to avoid the in-place operation. (code example above was updated, too)

**Comment by florian-huber** [comment]
Sparse score computation and strorage has been implemented in #370.
(only aspect missing: `.matrix()` method needs full sparse consideration as well, but this is on the todo elsewhere).
